src/Main.elm

```diff
@@ -1,37 +1,143 @@
-module Main exposing (Model, Msg(..), book, init, main, update, view)
+module Main exposing (Model, Msg(..), init, main, update, view)
 
 import BookData exposing (Book, allBooks)
 import Browser
-import Html exposing (Html, div, h1, img, li, text, ul)
+import Html exposing (Html, button, div, h1, img, li, table, td, text, th, thead, tr, ul)
 import Html.Attributes exposing (alt, class, src)
+import Html.Events exposing (onClick)
+import List exposing (append, filter, map, sum)
 
 
 
 ---- MODEL ----
 
 
 type alias Model =
-    { books : List Book }
+    { books : List Book, cart : Cart }
+
+
+type alias CartItem =
+    { id : BookID, quantity : Int }
+
+
+type alias Cart =
+    List CartItem
+
+
+type alias BookID =
+    Int
+
+
+add : Cart -> BookID -> Cart
+add cart bookID =
+    -- let
+    --     _ =
+    --         Debug.log "Logging b" bookID
+    -- in
+    case inCart bookID cart of
+        True ->
+            let
+                incrementQuantity cartItem =
+                    if cartItem.id == bookID then
+                        { cartItem | quantity = cartItem.quantity + 1 }
+
+                    else
+                        cartItem
+            in
+            List.map (\item -> incrementQuantity item) cart
```

Comment:
This is a nice, but I think ultimately using `replaceIf` is preferable. It's possibly a bit misleading to have `incrementQuantity` as a map function when you're only sometimes incrementing the quantity.

Reply:
Just noting for myself that replaceIf is now named setIf in the most recent List.Extra

Reply:
Having difficulties figuring out how to use setIf to replace a value inside a record within the list - I can't write it like

`setIf (\cartItem -> cartItem.id == bookID) { cartItem | quantity = cartItem.quantity + 1 } cart`

(as per `setIf : (a -> Bool) -> a -> List a -> List a` )

as cartItem only exists inside the anonymous function; haven't quite figured out is there's a way to do it with `let ... in`